Android ONNX INT8 (NNAPI/ORT)

WangHHY19931001 opened the pull request:
call out that Ultralytics ONNX exports are NCHW while TFLite→ONNX often yields NHWC, and that class names can be read from the ONNX names metadata we write during export

<!--
Thank you 🙏 for your contribution to [Ultralytics](https://www.ultralytics.com/) 🚀! Your effort in enhancing our repositories is greatly appreciated. To streamline the process and assist us in integrating your Pull Request (PR) effectively, please follow these steps:

1. Check for Existing Contributions: Before submitting, kindly explore existing PRs to ensure your contribution is unique and complementary.
2. Link Related Issues: If your PR addresses an open issue, please link it in your submission. This helps us better understand the context and impact of your contribution.
3. Elaborate Your Changes: Clearly articulate the purpose of your PR. Whether it's a bug fix or a new feature, a detailed description aids in a smoother integration process.
4. Ultralytics Contributor License Agreement (CLA): To uphold the quality and integrity of our project, we require all contributors to sign the CLA. Please confirm your agreement by commenting below:

    I have read the CLA Document and I sign the CLA

For more detailed guidance and best practices on contributing, refer to our ✅ [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Your adherence to these guidelines ensures a faster and more effective review process.
-->

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Add Android example for running YOLO11 INT8 ONNX models via ONNX Runtime (with optional NNAPI), including build files, export scripts, and C++ inference pipeline.

### 📊 Key Changes
- New example: examples/YOLO11-CPP-ONNXRUNNER-INT8/ with CMake project and source files
- C++ INT8 ONNX inference implementation (Ort::Session, NHWC preprocessing, per-class NMS)
- TFLite INT8 export script (export_tflite.py) and TFLite→ONNX conversion metadata helper (change_onnx_mate.py)
- README with Android build and usage instructions

### 🎯 Purpose & Impact
- Provide a ready-to-build Android ONNX INT8 example leveraging ORT/NNAPI for efficient edge deployment
- Demonstrates full pipeline: train → TFLite INT8 export → ONNX conversion → Android inference
- Helps embedded developers accelerate YOLO11 on-device inference with minimal setup

Y-T-G committed:
Use PNNX Python interface for NCNN conversion (#22347)

Signed-off-by: Mohammed Yasin <32206511+Y-T-G@users.noreply.github.com>
Signed-off-by: Glenn Jocher <glenn.jocher@ultralytics.com>
Co-authored-by: UltralyticsAssistant <web@ultralytics.com>
Co-authored-by: Glenn Jocher <glenn.jocher@ultralytics.com>
Co-authored-by: Ultralytics Assistant <135830346+UltralyticsAssistant@users.noreply.github.com>
Co-authored-by: Laughing-q <1185102784@qq.com>
Co-authored-by: Jing Qiu <61612323+Laughing-q@users.noreply.github.com>
Signed-off-by: WangHHY19931001 <wanghhy@163.com>

WangHHY19931001 committed:
Android ONNX INT8 (NNAPI/ORT)
call out that Ultralytics ONNX exports are NCHW while TFLite→ONNX often yields NHWC, and that class names can be read from the ONNX names metadata we write during export

Signed-off-by: WangHHY19931001 <wanghhy@163.com>

WangHHY19931001 committed:
Android ONNX INT8 (NNAPI/ORT)
call out that Ultralytics ONNX exports are NCHW while TFLite→ONNX often yields NHWC, and that class names can be read from the ONNX names metadata we write during export

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

UltralyticsAssistant reviewed (commented):
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Rename the `and` parameter/variable to a valid identifier (e.g., `results`) throughout the declaration and definition so the project builds successfully.

💬 Posted 1 inline comment

WangHHY19931001 commented on examples/YOLO11-CPP-ONNXRUNNER-INT8/ott_check_for_int8.h:
```diff
@@ -0,0 +1,45 @@
+        return class_names;
+    }
+    // 注意input格式为BGR
+    bool Process(const cv::Mat& input, std::vector<OttCheckAns>& ans);
```
and in where?

WangHHY19931001 committed:
Run ruff check \
RUF003 Comment contains ambiguous `，` (FULLWIDTH COMMA). Did you mean `,` (COMMA)?
  --> examples/YOLO11-CPP-ONNXRUNNER-INT8/export_tflite.py:16:36
   |
14 |     nms=False,
15 |     fraction=1,
16 |     device="cpu",  # 指定导出设备为CPU或GPU，对应参数为"cpu" , "0"
   |                                            ^^
17 | )
   |

Found 1 error.
Error: Process completed with exit code 1.

fix it

Signed-off-by: WangHHY19931001 <wanghhy@163.com>

WangHHY19931001 committed:
Run ruff check \
RUF003 Comment contains ambiguous `，` (FULLWIDTH COMMA). Did you mean `,` (COMMA)?
  --> examples/YOLO11-CPP-ONNXRUNNER-INT8/export_tflite.py:16:36
   |
14 |     nms=False,
15 |     fraction=1,
16 |     device="cpu",  # 指定导出设备为CPU或GPU，对应参数为"cpu" , "0"
   |                                            ^^
17 | )
   |

Found 1 error.
Error: Process completed with exit code 1.

fix it

WangHHY19931001 committed:
Run ruff check \
RUF003 Comment contains ambiguous `，` (FULLWIDTH COMMA). Did you mean `,` (COMMA)?
  --> examples/YOLO11-CPP-ONNXRUNNER-INT8/export_tflite.py:16:36
   |
14 |     nms=False,
15 |     fraction=1,
16 |     device="cpu",  # 指定导出设备为CPU或GPU，对应参数为"cpu" , "0"
   |                                            ^^
17 | )
   |

Found 1 error.
Error: Process completed with exit code 1.

fixit

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

WangHHY19931001 committed:
I have read the CLA Document and I hereby sign the CLA

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

WangHHY19931001 committed:
add build env

Signed-off-by: WangHHY19931001 <wanghhy@163.com>

WangHHY19931001 committed:
Auto-format by https://ultralytics.com/actions

Signed-off-by: WangHHY19931001 <wanghhy@163.com>

WangHHY19931001 committed:
Merge branch 'main' into main

WangHHY19931001 commented on examples/YOLO11-CPP-ONNXRUNNER-INT8/README.md:
```diff
@@ -0,0 +1,25 @@
+# Android ONNX INT8 (NNAPI/ORT)
```
I have read the CLA Document and I hereby sign the CLA

WangHHY19931001 commented on examples/YOLO11-CPP-ONNXRUNNER-INT8/change_onnx_mate.py:
```diff
@@ -0,0 +1,54 @@
+import onnx
```
I have read the CLA Document and I hereby sign the CLA

WangHHY19931001 commented on examples/YOLO11-CPP-ONNXRUNNER-INT8/ott_check_for_int8.h:
```diff
@@ -0,0 +1,45 @@
+//
```
I have read the CLA Document and I hereby sign the CLA

WangHHY19931001 reviewed (commented):
I have read the CLA Document and I hereby sign the CLA

WangHHY19931001 committed:
del all and upload android arm64 onnxruntime yolo int8 demo

Signed-off-by: WangHHY19931001 <wanghhy@163.com>

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

WangHHY19931001 committed:
Merge branch 'main' into main

onuralpszr commented:
@WangHHY19931001 : https://github.com/ultralytics/ultralytics/blob/69527c280e799ca71b8827be604ac6720f2c3072/examples/YOLO11-CPP-ONNXRUNNER-INT8/app/src/main/assets/test_2400_1080.png 

Before you deleting this let me ask you this, did you win at least ? :)

WangHHY19931001 commented:
> @WangHHY19931001 : https://github.com/ultralytics/ultralytics/blob/69527c280e799ca71b8827be604ac6720f2c3072/examples/YOLO11-CPP-ONNXRUNNER-INT8/app/src/main/assets/test_2400_1080.png
> 
> Before you deleting this let me ask you this, did you win at least ? :)

it work

<img width="1300" height="881" alt="image" src="https://github.com/user-attachments/assets/679cb04f-1e45-4172-b797-81e7bedcbd24" />

WangHHY19931001 committed:
1. minsize test image
2. add image copyright
3. change value name

Signed-off-by: WangHHY19931001 <wanghhy@163.com>

WangHHY19931001 committed:
Merge branch 'main' into main

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

WangHHY19931001 committed:
Merge branch 'main' into main

WangHHY19931001 committed:
Merge branch 'ultralytics:main' into main
